### ISSUE TITLE
Uniformity annotations for global variables

### ISSUE THREAD

**Issue by RobinMorisset** [gpuweb/gpuweb#1791]
The uniformity analysis (still under discussion, see https://github.com/gpuweb/gpuweb/pull/1571) treats all accesses to global variables as non-uniform.

If I understood @kvark right, Naga currently uses global variables that always contain uniform data when compiling some builtins from SPIRV. It would be quite easy to support this use case by providing an annotation that could be added to these variables.

It would be trivial for the analysis to use these annotations:
- Verify that any store to such a variable is of a uniform value (and in uniform control-flow)
- Assume that every read from such a variable sees a uniform value.

Such annotations may also be useful for programmers outside of Naga, so I'd rather have them as something supported in the spec than as a Naga-specific hack.

I have no opinion on the syntax that would be best for them. The simplest might be an attribute [AlwaysUniformValue] or something like that.

**Comment by kvark** [comment]
Thank you! Note that Tint does the same.

**Comment by jdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2021-06-08-Minutes#uniformity-annotations-for-global-variables-1791">WGSL meeting minutes 2021-06-08</a></summary>



*   DN: Can we have the compiler just figure out uniform/non-uniform for the vars, rather than needing the author to get it right?
*   RM: The alg can only validate yes/no, so it would need to try yes/no for each var, so it’d be exponential in runtime.
*   MM: Why not model the vars as inputs (and outputs?) to functions?
*   RM: Same sort of runtime problem, where complexity is higher that way.
*   RM: Not sure what syntax to propose here.
*   DN: Suggest making a new PR for this thing alone. Also, “uniform” at what scope? Especially once we have subgroups.
*   MM: Sounds like we want this between double-square-brackets.


</details>

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-05-17-minutes#uniformity-annotations-for-global-variables--issue-1791">WGSL meeting minutes 2022-05-17</a></summary>



* AB: Proposed post-v1 or post uniformity feedback which is starting to trickle in. Think wait on feedback. We have a label for uniformity and this is tagged.
* DN: Currently resolved/needs-spec should be post-mvp.
* KG: **Mark it polish post-v1**.
* AB: Think a lot of this comes from consideration of translating from SPIR-V -> WGSL. Not a v1 target for us.


</details>

**Comment by alan-baker** [comment]
I expect this is no longer desirable given where we ended up with diagnostics. It would also be awkward to annotate a variable as uniform only for compute shaders to not respect that annotation.

**Comment by Kangz** [comment]
Compute shaders have to respect the uniformity analysis though?

**Comment by alan-baker** [comment]
> Compute shaders have to respect the uniformity analysis though?

Yes when we did the diagnostics and uniformity opt out, you cannot opt out of the analysis from control barriers in compute shaders.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2023-11-28-minutes#uniformity-annotations-for-global-variables--issue-1791">WGSL 2023-11-28 Minutes</a></summary>

* AB: Think we can close this. We added a diagnostic system.  That’s what this was originally proposed for.
* JB: What about compute shaders.
* AB: If it’s unsafe to turn off, then you can’t trust the user to tell you to turn it off anyway.
* JB: The user is making a promise. For private variables, you should trust the user. If it is sound to forbid non-uniform writes, then why can’t we infer that uniformity.
* JB: Don’t know if Robin’s original proposal is correct.
* KG: We can close this for now, then wait for folks to complain. Then we can reopen it.
* **_Close it._**


---

</details>

**Comment by kainino0x** [comment]
+1, I think the proposal that's being rejected in the meeting minutes is "escape hatch to assert uniformity of global variables", but the proposal in the OP is "introduce another type of global variable that is checked to be uniform"

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2024‐01‐09-Minutes#uniformity-annotations-for-global-variables--issue-1791">WGSL 2024-01-09 Minutes</a></summary>

* AB: We’re confused.  I don’t understand why it was reopened.  Don’t think it’s necessary.
* JB: In terms of making an assertion about the variable, “treat this as uniform”; we didn’t think that was value.  It’s nice for the user to state their assumption and make a request to check; then they can treat as an error or warning. It wouldn’t change the conclusions of the analysis.
* BC: That’s nice, but would be hesitant to add more controls to the uniformity analysis until someone requests it in a real example. So far what I’ve seen is confusion on the part of users. 
* JB: This is not more control. It’s helpful for people who are confused.  You can annotate intermediates and have it checked.  Strictly gives more information about what the uniformity analysis thinks.
* AB: It’s not exactly. It’s an assumption currently that writable global variables are non-uniform.  So now you’re adding more endpoints to the check to verify. It’s not free because it’s not something that’s implemented.
* BC: Most of the folks I’ve been helping aren’t even sure what uniformity means. There’s a certain level of learning to be done before we give more tools.
* JB: Platforms do care what uniformity is. 
* DN: Sounds like what you want is uniform(X) that you can then static-assert on.
* JP: That’s a different use case than what is being requested here.  E.g. tf.js ran into this; we had to change Tint’s output to accommodate.
* DN: Still don’t think we’re in consensus that this is good to add.
* AB: Happy to push to M3+, We’ve had solutions for folks so far. Don’t think there’s a pressing need.
* JP: Clients that run across this in their own code can pass values down as parameters.
* **_Move to M3+_**

</details>
